url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/core/CloudSim.java

```diff
@@ -0,0 +1,928 @@
+						if ((p == null) || (tag == 9999) || (p.match(e))) {
+							dest_ent.setEventBuffer((SimEvent) e.clone());
+							dest_ent.setState(SimEntity.RUNNABLE);
+							waitPredicates.remove(destObj);
+						} else {
+							deferred.addEvent(e);
+						}
+					} else {
+						deferred.addEvent(e);
+					}
+				}
+				break;
+
+			case SimEvent.HOLD_DONE:
+				src = e.getSource();
+				if (src < 0) {
+					throw new IllegalArgumentException("Null entity holding.");
+				} else {
+					entities.get(src).setState(SimEntity.RUNNABLE);
+				}
+				break;
+
+			default:
+				break;
+		}
+	}
+
+	/**
+	 * Internal method used to start the simulation. This method should <b>not</b> be used by user
+	 * simulations.
+	 */
+	public static void runStart() {
+		running = true;
+		// Start all the entities
+		for (SimEntity ent : entities) {
+			ent.startEntity();
+		}
+
+		printMessage("Entities started.");
+	}
+
+	/**
+	 * Check if the simulation is still running. This method should be used by entities to check if
+	 * they should continue executing.
+	 * 
+	 * @return if the simulation is still running, otherwise
+	 */
+	public static boolean running() {
+		return running;
+	}
+
+	/**
+	 * This method is called if one wants to pause the simulation.
+	 * 
+	 * @return true, if successful otherwise.
+	 */
+	public static boolean pauseSimulation() {
+		paused = true;
+		return paused;
+	}
+
+	/**
+	 * This method is called if one wants to pause the simulation at a given time.
+	 * 
+	 * @param time the time at which the simulation has to be paused
+	 * @return true, if successful otherwise.
+	 */
+	public static boolean pauseSimulation(long time) {
+		if (time <= clock) {
+			return false;
+		} else {
+			pauseAt = time;
+		}
+		return true;
+	}
+
+	/**
+	 * This method is called if one wants to resume the simulation that has previously been paused.
+	 * 
+	 * @return if the simulation has been restarted or or otherwise.
+	 */
+	public static boolean resumeSimulation() {
+		paused = false;
+
+		if (pauseAt <= clock) {
+			pauseAt = -1;
+		}
+
+		return !paused;
+	}
+
+	/**
+	 * Start the simulation running. This should be called after all the entities have been setup
+	 * and added, and their ports linked.
+	 * 
+	 * @return the double last clock value
+	 */
+	public static double run() {
+		if (!running) {
+			runStart();
+		}
+		while (true) {
+			if (runClockTick() || abruptTerminate) {
+				break;
+			}
+
+			// this block allows termination of simulation at a specific time
+			if (terminateAt > 0.0 && clock >= terminateAt) {
+				terminateSimulation();
+				clock = terminateAt;
+				break;
+			}
+
+			if (pauseAt != -1
+					&& ((future.size() > 0 && clock <= pauseAt && pauseAt <= future.iterator().next()
+							.eventTime()) || future.size() == 0 && pauseAt <= clock)) {
+				pauseSimulation();
+				clock = pauseAt;
+			}
+
+			while (paused) {
+				try {
+					Thread.sleep(100);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)